src/devices/bus/mononcol/slot.h

```diff
@@ -0,0 +1,194 @@
+
+	// slot interface overrides
+	virtual std::string get_default_card_software(get_default_card_software_hook &hook) const override;
+
+	u32 common_get_size(char const *region);
+	void common_load_rom(u8 *ROM, u32 len, char const *region);
+
+	virtual void rom_alloc(u32 size, int width, endianness_t end) { if (m_cart) m_cart->rom_alloc(size, width, end, tag()); }
+
+	u8* get_rom_base()
+	{
+		if (!m_cart)
+			return nullptr;
+		else if (!user_loadable())
+			return m_cart->get_region_base();
+		else
+			return m_cart->get_rom_base();
+	}
+	u32 get_rom_size()
+	{
+		if (!m_cart)
+			return 0U;
+		else if (!user_loadable())
+			return m_cart->get_region_size();
+		else
+			return m_cart->get_rom_size();
+	}
+
+	uint8_t read()
+	{
+		return m_cart->read();
+	}
+
+	DECLARE_WRITE_LINE_MEMBER(dir_w)
+	{
+		m_cart->dir_w(state);
+	}
+
+	void set_ready()
+	{
+		m_cart->set_ready();
+	}
+
+	void write(uint8_t data)
+	{
+		m_cart->write(data);
+	}
+
+	void set_spi_region(uint8_t* region) { m_cart->set_spi_region(region); }
+	void set_spi_size(size_t size) { m_cart->set_spi_size(size); }
+
+
+protected:
+	mononcol_slot_device(machine_config const &mconfig, device_type type, char const *tag, device_t *owner, u32 clock);
+
+	// device-level overrides
+	virtual void device_start() override ATTR_COLD;
+
+	char const *m_interface;
+	char const *m_default_card;
+	char const *m_extensions;
+	int m_width;
+	endianness_t m_endianness;
+	device_mononcol_cart_interface *m_cart;
+	load_delegate m_device_image_load;
+	unload_delegate m_device_image_unload;
+};
+
+class mononcol_socket_device : public mononcol_slot_device
+{
+public:
+	template <typename T>
+	mononcol_socket_device(machine_config const &mconfig, char const *tag, device_t *owner, T &&opts, char const *intf, char const *exts = nullptr)
+		: mononcol_socket_device(mconfig, tag, owner, u32(0))
+	{
+		opts(*this);
+		set_fixed(false);
+		set_interface(intf);
+		if (exts)
+			set_extensions(exts);
+	}
+
+	mononcol_socket_device(machine_config const &mconfig, char const *tag, device_t *owner, u32 clock = 0);
+
+	virtual const char *image_type_name() const noexcept override { return "romimage"; }
+	virtual const char *image_brief_type_name() const noexcept override { return "rom"; }
+};
+
+class mononcol_cartslot_device : public mononcol_slot_device
+{
+public:
+	template <typename T>
+	mononcol_cartslot_device(machine_config const &mconfig, char const *tag, device_t *owner, T &&opts, char const *intf, char const *exts = nullptr)
+		: mononcol_cartslot_device(mconfig, tag, owner, u32(0))
+	{
+		opts(*this);
+		set_fixed(false);
+		set_interface(intf);
+		if (exts)
+			set_extensions(exts);
+	}
+
+	mononcol_cartslot_device(machine_config const &mconfig, char const *tag, device_t *owner, u32 clock = 0);
+
+	virtual const char *image_type_name() const noexcept override { return "cartridge"; }
+	virtual const char *image_brief_type_name() const noexcept override { return "cart"; }
+};
```

Comment:
Are these derived classes actually necessary?  You only have a cartridge slot, right?  You shoudn’t need the “ROM socket” device.

Reply:
probably not, I just cloned the generic stuff to get up and running with this, so some of it can be reduced down further